editor/editor_undo_redo_manager.cpp

```diff
@@ -0,0 +1,439 @@
+		{
+			History &history = get_or_create_history(GLOBAL_HISTORY);
+			if (!history.undo_stack.is_empty()) {
+				selected_history = &history;
+				global_timestamp = history.undo_stack.back()->get().timestamp;
+			}
+		}
+
+		{
+			History &history = get_or_create_history(EditorNode::get_editor_data().get_current_edited_scene_history_id());
+			if (!history.undo_stack.is_empty() && history.undo_stack.back()->get().timestamp > global_timestamp) {
+				selected_history = &history;
+			}
+		}
+
+		if (selected_history) {
+			return selected_history->undo_redo->get_current_action_name();
+		}
+	}
+	return "";
+}
+
+void EditorUndoRedoManager::discard_history(int p_idx) {
+	ERR_FAIL_COND(!history_map.has(p_idx));
+	History &history = history_map[p_idx];
+
+	if (history.undo_redo) {
+		memdelete(history.undo_redo);
+		history.undo_redo = nullptr;
+	}
+	history_map.erase(p_idx);
+}
+
+void EditorUndoRedoManager::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("create_action", "name", "merge_mode", "custom_context"), &EditorUndoRedoManager::create_action, DEFVAL(UndoRedo::MERGE_DISABLE), DEFVAL((Object *)nullptr));
+	ClassDB::bind_method(D_METHOD("commit_action", "execute"), &EditorUndoRedoManager::commit_action, DEFVAL(true));
+	ClassDB::bind_method(D_METHOD("is_committing_action"), &EditorUndoRedoManager::is_committing_action);
+
+	{
+		MethodInfo mi;
+		mi.name = "add_do_method";
+		mi.arguments.push_back(PropertyInfo(Variant::OBJECT, "object"));
+		mi.arguments.push_back(PropertyInfo(Variant::STRING_NAME, "method"));
+
+		ClassDB::bind_vararg_method(METHOD_FLAGS_DEFAULT, "add_do_method", &EditorUndoRedoManager::_add_do_method, mi, varray(), false);
+	}
+
+	{
+		MethodInfo mi;
+		mi.name = "add_undo_method";
+		mi.arguments.push_back(PropertyInfo(Variant::OBJECT, "object"));
+		mi.arguments.push_back(PropertyInfo(Variant::STRING_NAME, "method"));
+
+		ClassDB::bind_vararg_method(METHOD_FLAGS_DEFAULT, "add_undo_method", &EditorUndoRedoManager::_add_undo_method, mi, varray(), false);
+	}
+
+	ClassDB::bind_method(D_METHOD("add_do_property", "object", "property", "value"), &EditorUndoRedoManager::add_do_property);
+	ClassDB::bind_method(D_METHOD("add_undo_property", "object", "property", "value"), &EditorUndoRedoManager::add_undo_property);
+	ClassDB::bind_method(D_METHOD("add_do_reference", "object"), &EditorUndoRedoManager::add_do_reference);
+	ClassDB::bind_method(D_METHOD("add_undo_reference", "object"), &EditorUndoRedoManager::add_undo_reference);
+
+	ClassDB::bind_method(D_METHOD("get_object_history_id", "object"), &EditorUndoRedoManager::get_history_id_for_object);
+	ClassDB::bind_method(D_METHOD("get_history_undo_redo", "object"), &EditorUndoRedoManager::get_history_undo_redo);
+
+	BIND_ENUM_CONSTANT(GLOBAL_HISTORY);
+	BIND_ENUM_CONSTANT(INVALID_HISTORY);
+}
+
+EditorUndoRedoManager::~EditorUndoRedoManager() {
+	for (const KeyValue<int, History> &E : history_map) {
+		discard_history(E.key);
+	}
+}
```

Comment:
Sanitizer find here use after free, so maybe during iterating over this collection, its elements are removed.

Iterating on copy of history_map or using while maybe should fix issue:
```
while (history_map.length() != 0) {
discard_history.pop()
}
```